scripts/generate_changelog.rb

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'net/http'
+require 'json'
+require 'open3'
+
+# Pretty format for git logs. Outputs all commit notes delimited by the @@ symbol.
+#   @@
+#   @feature
+#   Note content
+#   @@
+#   @bug
+#   Note Content
+FORMAT = "@@%n%N"
+
+# Parses individual notes by searching for a @label, normalizing it to a
+# valid entry key, and returning both the normalized label and body of the note
+def parse_note(labels, note)
+  label, lines = note.split("\n").partition do |line|
+    line.match?(/\A@(#{labels.join('|')})\z/)
+  end
+
+  entry = lines.join("\n").strip
+
+  case label.count
+  when 0
+    warn "Did not find label for note:\n#{entry}"
+    exit 1
+  when 1
+    label = label.first[1..-1]
+  else
+    warn "Found multiple labels for note:\n#{entry}"
+    exit 1
+  end
+
+  [label, entry]
+end
+
+# Map of valid labels to their full name and list of entries.
+# Any notes that do not have a valid label are ignored and not added
+# to the changelog.
+entries = {
+  'feature'     => { name: 'New features', entries: [] },
+  'bug'         => { name: 'Bug fixes',    entries: [] },
+  'deprecation' => { name: 'Deprecations', entries: [] },
+  'removal'     => { name: 'Removals',     entries: [] }
+}
+
+version    = ARGV.first
+changelog  = File.expand_path('CHANGELOG.md', Dir.pwd)
+
+unless version
+  warn "Usage: generate_changelog.rb VERSION"
+  exit 1
+end
+
+unless File.file?(changelog)
+  warn "Could not find changelog at #{changelog}"
+  exit 1
+end
+
+# Request the latest version of Bolt from GitHub
+uri    = URI('https://api.github.com/repos/puppetlabs/bolt/tags')
+result = Net::HTTP.get(uri)
+latest = JSON.parse(result).first['name']
+
+# Read the git log between master and the most recent tagged release
+# The log is formatted to only include notes, with each entry demilited
+# by @@. These entries are then filtered to only include those that have
+# a valid LABEL (e.g. @feature, @bug)
+stdout_str, stderr_str, status = Open3.capture3(
+  "git log master...#{latest} "\
+  "--pretty=format:'#{FORMAT}' "\
+  "--grep='@(#{entries.keys.join('|')})' -E"
+)
+
+if status.success?
+  # Create an array of notes, tossing the first one which will always be blank.
+  # TODO: There's probably a better way to do this so we don't get a
```

Comment:
Try `git log -z`. That will separate the entries with a `"\x00"` character instead of a newline, so you don't need to define your own marker (since a commit message won't have a null byte ordinarily).

Reply:
You should also be able to get rid of the `%n` part of the pretty format